firedrake/bcs.py

```diff
@@ -94,6 +84,201 @@ def __init__(self, V, g, sub_domain, method="topological"):
+                bcnodes1 = []
+                if len(s) > 1 and not isinstance(self._function_space.finat_element, finat.Lagrange):
+                    raise TypeError("Currently, edge conditions have only been tested with Lagrange elements")
+                for ss in s:
+                    # intersection of facets
+                    # Edge conditions have only been tested with Lagrange elements.
+                    # Need to expand the list.
+                    bcnodes1.append(hermite_stride(self._function_space.boundary_nodes(ss, self.method)))
+                bcnodes1 = functools.reduce(np.intersect1d, bcnodes1)
+                bcnodes.append(bcnodes1)
+            return np.concatenate(tuple(bcnodes))
+
+    @utils.cached_property
+    def node_set(self):
+        '''The subset corresponding to the nodes at which this
+        boundary condition applies.'''
+
+        return op2.Subset(self._function_space.node_set, self.nodes)
+
+    def zero(self, r):
+        r"""Zero the boundary condition nodes on ``r``.
+
+        :arg r: a :class:`.Function` to which the
+            boundary condition should be applied.
+
+        """
+        if isinstance(r, matrix.MatrixBase):
+            raise NotImplementedError("Zeroing bcs on a Matrix is not supported")
+
+        for idx in self._indices:
+            r = r.sub(idx)
+        try:
+            r.dat.zero(subset=self.node_set)
+        except exceptions.MapValueError:
+            raise RuntimeError("%r defined on incompatible FunctionSpace!" % r)
+
+    def set(self, r, val):
+        r"""Set the boundary nodes to a prescribed (external) value.
+        :arg r: the :class:`Function` to which the value should be applied.
+        :arg val: the prescribed value.
+        """
+        for idx in self._indices:
+            r = r.sub(idx)
+            val = val.sub(idx)
+        r.assign(val, subset=self.node_set)
+
+    def integrals(self):
+        raise NotImplementedError("integrals() method has to be overwritten")
+
+    def as_subspace(self, field, V, use_split):
+        fs = self._function_space
+        if fs.parent is not None and isinstance(fs.parent.ufl_element(), VectorElement):
+            index = fs.parent.index
+        else:
+            index = fs.index
+        cmpt = fs.component
+        # TODO: need to test this logic
+        field_renumbering = dict([f, i] for i, f in enumerate(field))
+        if index in field:
+            if len(field) == 1:
+                W = V
+            else:
+                W = V.split()[field_renumbering[index]] if use_split else V.sub(field_renumbering[index])
+            if cmpt is not None:
+                W = W.sub(cmpt)
+            return W
+
+    def sorted_equation_bcs(self):
+        return []
+
+    def increment_bc_depth(self):
+        # Increment _bc_depth by 1
+        self._bc_depth += 1
+        for bc in itertools.chain(*self.bcs):
+            bc._bc_depth += 1
+
+
+class DirichletBC(BCBase):
```

Comment:
Is DirichletBC not a special case of EquationBCSplit when the operator is identity ? I do not understand why is there a need of a specific class for Dirichlet BC.

Is it to avoid solving the trivial equation u = BCoperator(u) = g ?

Reply:
Effectively yes. We can apply certain other optimisations (e.g. maintaining symmetry) in the DirichletBC case.